File: examples/go/servers/advanced/dynamic-pay-to.go
package main

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"time"

	ginfw "github.com/gin-gonic/gin"
	"github.com/joho/godotenv"
	x402 "github.com/x402-foundation/x402/go"
	x402http "github.com/x402-foundation/x402/go/http"
	ginmw "github.com/x402-foundation/x402/go/http/gin"
	evm "github.com/x402-foundation/x402/go/mechanisms/evm/exact/server"
)

const DefaultPort = "4021"

/**
 * Dynamic PayTo Example
 *
 * This example demonstrates how to use dynamic payTo resolution to
 * route payments to different addresses based on the request context.
 * This is useful for marketplace applications where payments should
 * go to different sellers/creators.
 */

func main() {
	godotenv.Load()

	evmPayeeAddress := os.Getenv("EVM_PAYEE_ADDRESS")
	if evmPayeeAddress == "" {
		fmt.Println("❌ EVM_PAYEE_ADDRESS environment variable is required")
		os.Exit(1)
	}

	facilitatorURL := os.Getenv("FACILITATOR_URL")
	if facilitatorURL == "" {
		fmt.Println("❌ FACILITATOR_URL environment variable is required")
		os.Exit(1)
	}

	evmNetwork := x402.Network("eip155:84532") // Base Sepolia

	r := ginfw.Default()

	facilitatorClient := x402http.NewHTTPFacilitatorClient(&x402http.FacilitatorConfig{
		URL: facilitatorURL,
	})

	/**
	 * Address Lookup Table
	 *
	 * In a real application, you would query a database for the appropriate
	 * payee address. Here we use a simple lookup table for demonstration.
	 */
	addressLookup := map[string]string{
		"US": evmPayeeAddress,
		"UK": evmPayeeAddress,
		"CA": evmPayeeAddress,
		"AU": evmPayeeAddress,
		"NZ": evmPayeeAddress,
		"IE": evmPayeeAddress,
		"FR": evmPayeeAddress,
	}

	/**
	 * Dynamic PayTo Function
	 *
	 * This function is called at request time to determine where the payment
	 * should be sent. It receives the full HTTP request context.
	 */
	dynamicPayTo := func(ctx context.Context, reqCtx x402http.HTTPRequestContext) (string, error) {
		// Get the country from query parameter
		country := "US" // default
		if reqCtx.Adapter != nil {
			// In a real implementation, you'd extract query params from the adapter
			// For now, we'll use the default
		}

		address, ok := addressLookup[country]
		if !ok {
			address = evmPayeeAddress // fallback to default
		}

		fmt.Printf("💰 Routing payment for country=%s to address=%s\n", country, address)
		return address, nil
	}

	routes := x402http.RoutesConfig{
		"GET /weather": {
			Accepts: x402http.PaymentOptions{
				{
					Scheme:  "exact",
					PayTo:   x402http.DynamicPayToFunc(dynamicPayTo),
					Price:   "$0.001",
					Network: evmNetwork,
				},
			},
			Description: "Weather data",
			MimeType:    "application/json",
		},
	}

	r.Use(ginmw.X402Payment(ginmw.Config{
		Routes:      routes,
		Facilitator: facilitatorClient,
		Schemes: []ginmw.SchemeConfig{
			{Network: evmNetwork, Server: evm.NewExactEvmScheme()},
		},
		SyncFacilitatorOnStart: true,
		Timeout:                30 * time.Second,
	}))

	r.GET("/weather", func(c *ginfw.Context) {
		c.JSON(http.StatusOK, ginfw.H{
			"report": ginfw.H{
				"weather":     "sunny",
				"temperature": 70,
			},
		})
	})

	fmt.Printf("🚀 Dynamic PayTo example running on http://localhost:%s\n", DefaultPort)
	fmt.Printf("   Payments are routed based on request context\n")

	if err := r.Run(":" + DefaultPort); err != nil {
		fmt.Printf("Error starting server: %v\n", err)
		os.Exit(1)
	}
}

File: examples/python/servers/advanced/dynamic_pay_to.py
"""Dynamic pay-to routing example."""

import os

from dotenv import load_dotenv
from fastapi import FastAPI
from pydantic import BaseModel

from x402.http import FacilitatorConfig, HTTPFacilitatorClient, PaymentOption
from x402.http.middleware.fastapi import PaymentMiddlewareASGI
from x402.http.types import HTTPRequestContext, RouteConfig
from x402.mechanisms.evm.exact import ExactEvmServerScheme
from x402.schemas import Network
from x402.server import x402ResourceServer

load_dotenv()

# Config
EVM_ADDRESS = os.getenv("EVM_ADDRESS")
EVM_NETWORK: Network = "eip155:84532"  # Base Sepolia
FACILITATOR_URL = os.getenv("FACILITATOR_URL", "https://x402.org/facilitator")

if not EVM_ADDRESS:
    raise ValueError("Missing required EVM_ADDRESS environment variable")

# Address lookup for dynamic pay-to
ADDRESS_LOOKUP: dict[str, str] = {
    "US": EVM_ADDRESS,
    "UK": EVM_ADDRESS,
    "CA": EVM_ADDRESS,
    "AU": EVM_ADDRESS,
}


def get_dynamic_pay_to(context: HTTPRequestContext) -> str:
    """Get dynamic pay-to address based on country query parameter."""
    country = context.adapter.get_query_param("country") or "US"
    return ADDRESS_LOOKUP.get(country, EVM_ADDRESS)


class WeatherReport(BaseModel):
    weather: str
    temperature: int


class WeatherResponse(BaseModel):
    report: WeatherReport


app = FastAPI()

facilitator = HTTPFacilitatorClient(FacilitatorConfig(url=FACILITATOR_URL))
server = x402ResourceServer(facilitator)
server.register(EVM_NETWORK, ExactEvmServerScheme())


# Register hooks to log selected payment option
async def after_verify(ctx):
    print("\n=== Dynamic Pay-To - After verify ===")
    print(f"Pay to: {ctx.requirements.pay_to}")
    print(f"Payer: {ctx.result.payer}")


server.on_after_verify(after_verify)

routes = {
    "GET /weather": RouteConfig(
        accepts=[
            PaymentOption(
                scheme="exact",
                pay_to=get_dynamic_pay_to,
                price="$0.001",
                network=EVM_NETWORK,
            ),
        ],
    ),
}
app.add_middleware(PaymentMiddlewareASGI, routes=routes, server=server)


@app.get("/weather")
async def get_weather(city: str = "San Francisco", country: str = "US") -> WeatherResponse:
    return WeatherResponse(report=WeatherReport(weather="sunny", temperature=70))


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=4021)
